src/d3a/models/area/markets.py

```diff
@@ -15,84 +15,175 @@
+        else:
+            if ConstSettings.GeneralSettings.ENABLE_SETTLEMENT_MARKETS:
+                return (time_slot < current_time_slot.subtract(
+                    hours=ConstSettings.GeneralSettings.MAX_AGE_SETTLEMENT_MARKET_HOURS))
+            else:
+                return time_slot < current_time_slot.subtract(
+                    minutes=GlobalConfig.slot_length.total_minutes())
+
+    def _is_it_time_to_rotate_market(self, current_time_slot: DateTime,
+                                     time_slot: DateTime) -> bool:
+        """Check if it is time to move market for time_slot into past markets."""
+        return time_slot < current_time_slot
+
+    def _move_markets_to_past(self, markets: Dict, past_markets: Dict,
+                              current_time_slot: DateTime) -> None:
+        """Move markets to self.past_markets."""
+        move_market_slots = [
+            time_slot for time_slot in markets.keys()
+            if self._is_it_time_to_rotate_market(current_time_slot, time_slot)]
+        for time_slot in move_market_slots:
+            market = markets.pop(time_slot)
+            market.readonly = True
+            past_markets[time_slot] = market
+            log.debug("Moving %s to past.", past_markets[time_slot])
+
+    def _delete_past_markets(self, market_dict: Dict, current_time_slot: DateTime) -> None:
+        """Delete the unneeded markets from self.past_markets."""
+        delete_market_slots = [
+            time_slot for time_slot in market_dict.keys()
+            if self._is_it_time_to_delete_past_market(current_time_slot, time_slot)]
+        for time_slot in delete_market_slots:
+            self._delete_market_and_all_attributes(market_dict, time_slot)
+
+    @staticmethod
+    def _delete_market_and_all_attributes(market_buffer: Dict, time_slot: DateTime) -> None:
+        """Delete market and all its attributes from the provided market_buffer.
+        When searching for a memory leak, we found out, that garbage collector does not delete
+        all attributes of the markets correctly. One possible explanation is that these
+        attributes are also referenced between each other and on other places in which case
+        the GC will not delete it from the memory."""
+
+        if ConstSettings.GeneralSettings.EVENT_DISPATCHING_VIA_REDIS:
+            market_buffer[time_slot].redis_api.stop()
+        del market_buffer[time_slot].offers
+        del market_buffer[time_slot].trades
+        del market_buffer[time_slot].offer_history
+        del market_buffer[time_slot].notification_listeners
+        del market_buffer[time_slot].bids
+        del market_buffer[time_slot].bid_history
+        del market_buffer[time_slot].traded_energy
+        del market_buffer[time_slot]
+
+
+class SpotMarketRotator(MarketRotatorBase):
+    """Deal with market rotation of spot markets."""
+
+
+class BalancingMarketRotator(MarketRotatorBase):
+    """Deal with market rotation of balancing markets."""
```

Comment:
I would suggest removing these classes and directly use `MarketRotatorBase` (maybe even calling it `DefaultMarketRotator`).